crates/re_viewport/src/context_menu/mod.rs

```diff
@@ -69,232 +77,219 @@ pub fn context_menu_ui_for_item(
-            );
+        response
+    }
 
-            items.push(CloneSpaceViewItem::item(space_view_id));
+    // TODO(ab): return a `ListItem` to make those context menu nice to look at. This requires
+    // changes to the context menu UI code to support full-span highlighting.
+    /// Returns the label displayed by [`Self::ui`]'s default implementation.
+    fn label(&self, _ctx: &ContextMenuContext<'_>) -> String {
+        String::new()
+    }
 
-            items
-        }
-        SelectionSummary::ContentsItems(contents) => {
-            // exclude the root container from the list of contents, as it cannot be shown/hidden
-            // nor removed
-            let contents: Rc<Vec<_>> = Rc::new(
-                contents
-                    .into_iter()
-                    .filter(|c| {
-                        Some(*c) != viewport_blueprint.root_container.map(Contents::Container)
-                    })
-                    .collect(),
-            );
-
-            if contents.is_empty() {
-                vec![]
-            } else if let Some(root_container_id) = viewport_blueprint.root_container {
-                // The new container should be created in place of the right-clicked content, so we
-                // look for its parent and position, and fall back to the root container.
-                let clicked_content = match item {
-                    Item::Container(container_id) => Some(Contents::Container(*container_id)),
-                    Item::SpaceView(space_view_id) => Some(Contents::SpaceView(*space_view_id)),
-                    _ => None,
-                };
-                let (target_container_id, target_position) = clicked_content
-                    .and_then(|c| viewport_blueprint.find_parent_and_position_index(&c))
-                    .unwrap_or((root_container_id, 0));
-
-                vec![
-                    ContentVisibilityToggle::item(viewport_blueprint, contents.clone()),
-                    ContentRemove::item(contents.clone()),
-                    Separator::item(),
-                    SubMenu::item(
-                        "Move to new container",
-                        possible_child_container_kind(viewport_blueprint, target_container_id).map(
-                            |kind| {
-                                MoveContentsToNewContainer::item(
-                                    target_container_id,
-                                    target_position,
-                                    kind,
-                                    contents.clone(),
-                                )
-                            },
-                        ),
-                    ),
-                ]
-            } else {
-                vec![]
+    // ---
+
+    /// Process the provided [`Selection`].
+    ///
+    /// The default implementation dispatches to [`Self::process_store_id`] and friends.
+    fn process_selection(&self, ctx: &ContextMenuContext<'_>) {
```

Comment:
not entirely sold that protecting implementors from this match expression makes sense, find it a bit odd to have every call bottom out in a customizable call instead of just forcing the whole thing on everyone. Matter of taste I guess :shrug:

Reply:
A drawback of this new approach is that actions must match against `Item` twice now: once at `supports_selection` time, and once at `process_selection` time (actually, avoiding that was one of the primary driver for the initial design based on `summarise_selection`). So the rational for the default implementation of `process_selection` and the related `process_xxx` is basically to avoid writing those two (near identical) match statements in every single implementor.